Blockchain integrated

Assil10 opened the pull request:
---

## Pull Request — Blockchain Integration into Backend

###  Summary

This PR integrates the full blockchain investment flow into the backend of the Korpor platform. It establishes the connection between the off-chain logic and Ethereum smart contracts, ensuring transparency, traceability, and immutability for real estate investments. The Swagger documentation has also been fully implemented for all blockchain-related endpoints.

---

###  Key Features Added

- **Smart Contract Integration**
  - Connected to deployed contracts on Sepolia via Infura
  - On-chain `recordInvestment` logic for confirmed investments
  - Transaction hash (`tx_hash`) saved to DB for traceability

- **Paymee Callback Handling**
  - Validates payment amount
  - Records confirmed transactions to the blockchain
  - Updates `projects.current_amount`
  - Automatically flags `status = 'Funded'` when `goal_amount` is reached

- **New Routes & Controllers**
  - `/api/investments` – Manage blockchain investment creation
  - `/api/payment/callback` – Webhook integration with Paymee
  - `/api/rent/distribute` – Admin simulation for rent payouts
  - `/api/users/:walletAddress` – Profile includes investment + rent history

- **Swagger API Documentation**
  - Full Swagger/OpenAPI annotations for:
    - Investments
    - Payments
    - Projects
    - Rent
    - Users

- **Blockchain Health Monitor**
  - Added `/utils/healthCheck.js` to check:
    - Network status
    - Smart contract responsiveness
    - Recent TX hash confirmations

---
 
issue #24 
###  Tickets Completed

- ✅ PB036 — Investor asset security  
- ✅ PB037 — Admin transaction verify  
- ✅ PB038 — Immutable records  
- ✅ PB039 — Chain health monitoring  

---

###  Affected Areas

- `controllers/`
- `routes/`
- `services/paymee.service.js`
- `blockchain/investment.js`
- `artifacts/contracts/...`
- Swagger annotations in all routes
- New smart contract JSONs under `artifacts/`

---

###  Notes

- All blockchain logic is off-chain-triggered (users never interact with Web3).
- Payments still flow via Paymee in fiat (TND).
- Only confirmed payments are recorded on-chain and reflected in the database.

---



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Introduced full backend support for blockchain-based real estate investment, including smart contracts for investment management, project registry, and rent distribution.
  - Added API endpoints for managing projects, investments, payments, rent distributions, and user registration/profile retrieval.
  - Enabled integration with the Paymee payment gateway for fiat-based investment flows.
  - Implemented proportional rent distribution to investors and tracking of rent payout history.
  - Provided health check utilities for blockchain connectivity and contract responsiveness.

- **Documentation**
  - Added a comprehensive README detailing the blockchain workflow, technology stack, and architectural separation of on-chain and off-chain components.

- **Chores**
  - Updated environment configuration for blockchain and payment integration.
  - Added new database tables and indexes to support projects, investments, payments, users, and rent distributions.
  - Upgraded and added relevant dependencies for blockchain interaction and HTTP requests.

- **Refactor**
  - Enhanced database configuration to support both raw SQL and ORM queries.
  - Modularized payment and blockchain interaction logic for maintainability and clarity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ahmedjaziri31 commented on backend/src/utils/balance.js:
```diff
@@ -0,0 +1,13 @@
+const hre = require("hardhat");
+
+async function main() {
+    const provider = hre.ethers.provider;
+    const walletAddress = "0x62ef1b3BD681ec3716dD58A5bB6f6cC16bBFD743"; 
+    const balance = await provider.getBalance(walletAddress);
+    console.log("Wallet Balance:", hre.ethers.utils.formatEther(balance), "ETH");
+}
+
+main().catch((error) => {
+    console.error(error);
+    process.exitCode = 1;
+});
```
@Assil10 if u can refact

ahmedjaziri31 commented on backend/src/utils/balance.js:
```diff
@@ -0,0 +1,13 @@
+
+async function main() {
+    const provider = hre.ethers.provider;
+    const walletAddress = "0x62ef1b3BD681ec3716dD58A5bB6f6cC16bBFD743"; 
+    const balance = await provider.getBalance(walletAddress);
+    console.log("Wallet Balance:", hre.ethers.utils.formatEther(balance), "ETH");
```
look into this hardcoded waller address

ahmedjaziri31 commented on backend/src/blockchain/investment.js:
```diff
@@ -0,0 +1,30 @@
+  wallet
+);
+
+exports.recordInvestment = async (projectId, userAddress, amountTND) => {
+  try {
+    // Convert the amount (in TND) to an integer by multiplying by 100 (cents, for example)
+    const amountInCents = Math.round(amountTND * 100); // Now it's an integer without decimals
+
+    // Call the smart contract to record the investment (amount in TND)
+    const tx = await contract.recordInvestment(projectId, userAddress, amountInCents);
+    
+    // Wait for the transaction to be mined
+    const receipt = await tx.wait();
+
+    // Return the transaction hash (txHash) as confirmation
+    return receipt.transactionHash;
+  } catch (error) {
+    console.error("Error recording investment on blockchain:", error);
+    throw new Error("Blockchain transaction failed");
+  }
+};
```
you can look into those two points :
Improve error handling
Add transaction confirmation checks

this is very reccomended

ahmedjaziri31 commented on backend/src/blockchain/investment.js:
```diff
@@ -0,0 +1,30 @@
+const contractABI = require("../artifacts/contracts/InvestmentManager.sol/InvestmentManager.json").abi;
+require("dotenv").config();
+
+const provider = new ethers.providers.JsonRpcProvider(process.env.INFURA_SEPOLIA);
+const wallet = new ethers.Wallet(process.env.PRIVATE_KEY, provider);
+const contract = new ethers.Contract(
+  process.env.INVESTMENT_CONTRACT_ADDRESS,
+  contractABI,
+  wallet
+);
```
This review comment is just a recommandation to include the GCP blockchain api or KMS for a faster and more secure process 
u can just make this in consideration and not implement it only if we needed it in the future

Assil10 commented:
@ahmedjaziri31 


1. **Wallet Address in `balance.js`**: The hardcoded wallet address has now been replaced with a dynamic approach that fetches the wallet address from environment variables. This enhances the flexibility and security of the code, making it easier to manage different wallet addresses for various environments.


2. **Improvement in `recordInvestment` Function**: 
   - I have enhanced the error handling in the `recordInvestment` function. Now, it checks for invalid input parameters (such as projectId, userAddress, or amountTND) and throws appropriate error messages.
   - Additionally, I added **transaction confirmation checks**. The transaction now waits for 2 confirmations before proceeding, ensuring that the transaction is confirmed securely before making changes to the database.
   - I've also added more detailed error handling for network issues, gas estimation failures, and transaction timeouts, which should make the system more robust and resilient.

3. **GCP Blockchain AP / KMS Suggestion**: 
    For now, I will leave it as-is, but I’ll make sure to do more research on this option to assess its suitability for future integration if needed.


---

ahmedjaziri31 commented:
APIs test has been requested to close this ticket and mark it done in the tracker

ahmedjaziri31 commented:
api of transaction, api to check the transaction health (means to check if the transaction is saved in blockchain) or not, (u can add any api necessarily u can think of we can need in the future app)

Assil10 commented:
testing the healthCheck function :

we got the last confirmed transaction record on the blockchain and the related data about it 
![image](https://github.com/user-attachments/assets/e5159778-58a4-4277-a74d-8e3a0e710787)

we can compare it to the generated hash from the last transaction 
![image](https://github.com/user-attachments/assets/1001d81c-22a5-4007-9005-da26542c3ccb)

==>conclusion the record is successfully saved to the blockchain and db.

Assil10 commented:
@ahmedjaziri31  you can merge it and close this ticket and mark it done in the tracker

Assil10 commented:
@ahmedjaziri31  take a look

Assil10 commented:
@coderabbitai make the diagram again i need the interactions correct and without too much colors for better clarity
provide me with the plantUML also beside the diagram
